launch/launch/actions/reset_launch_configurations.py

```diff
@@ -79,9 +80,11 @@ def execute(self, context: LaunchContext):
         else:
             evaluated_configurations = {}
             for k, v in self.__launch_configurations.items():
-                evaluated_k = perform_substitutions(context, normalize_to_list_of_substitutions(k))
-                evaluated_v = perform_substitutions(context, normalize_to_list_of_substitutions(v))
-                evaluated_configurations[evaluated_k] = evaluated_v
+                if isinstance(k, Substitution) or isinstance(k, str):
+                    k = perform_substitutions(context, normalize_to_list_of_substitutions(k))
+                if isinstance(v, Substitution) or isinstance(v, str):
+                    v = perform_substitutions(context, normalize_to_list_of_substitutions(v))
+                evaluated_configurations[k] = v
```

Comment:
maybe we can keep it simpler like below?

```suggestion
                k = perform_substitutions(context, normalize_to_list_of_substitutions(k)) if isinstance(k, (Substitution, str)) else k
                v = perform_substitutions(context, normalize_to_list_of_substitutions(v)) if isinstance(v, (Substitution, str)) else v
                evaluated_configurations[k] = v
```

Reply:
The type of the value (`v`) we get here for ROS parameters is `Iterable[Tuple[str, str]]`, e.g., `[('foo', 'true')]`. Technically, `normalize_to_list_of_substitutions()` can handle `SomeSubstitutionsType`, which _can_ be an iterable/list: https://github.com/ros2/launch/blob/e1d12d595f2a7af7341fd685ea53ad302ea49d60/launch/launch/some_substitutions_type.py#L25-L30

Note that it can't be an iterable _of tuples_ like what we have here. However, this means that the current `isinstance()` check is too strict: lists of substitutions would not be substituted and concatenated correctly, since substitutions in iterables/lists are performed and then concatenated together.**\***

I think we have to check if it's an iterable where not all items are `Text` or `Path` or `Substitution`, in which case we do not substitute it. Otherwise, we let it eventually result in an error. Something like:

```suggestion
                # Move imports to the top of the file
                import collections.abc
                from pathlib import Path
                from ..substitution import Substitution

                def is_substitutable(value: Any) -> bool:
                    # Specifically look for iterables of non-substitutable values. Assume the rest
                    # is substitutable and let it error out if not
                    return (
                        not isinstance(value, collections.abc.Iterable)
                        or all(
                            isinstance(iter_value, (str, Path, Substitution))
                            for iter_value in value
                        )
                    )

                k = (
                    perform_substitutions(context, normalize_to_list_of_substitutions(k))
                    if is_substitutable(k) else k
                )
                v = (
                    perform_substitutions(context, normalize_to_list_of_substitutions(v))
                    if is_substitutable(v) else v
                )
                evaluated_configurations[k] = v
```

Not a totally robust solution, but it has a fairly small scope. It seems to work with the test I suggested as well as the provided reproducer, which I slightly modified to print the parameter value:

```py
from launch import LaunchDescription
from launch.actions import TimerAction
from launch.actions import LogInfo
from launch_ros.actions import SetParameter
from launch_ros.substitutions import Parameter

def generate_launch_description():

    return LaunchDescription(
        [
            SetParameter(name="foo", value="true"),
            TimerAction(
                period=0.5,
                actions=[
                    # Outputs:
                    #   Parameter foo=true
                    LogInfo(msg=['Parameter foo=', Parameter(name='foo')]),
                ],
            ),
        ]
    )
```

----

<details><summary>(*) click to expand</summary>
<p>

This explains the test failures in the Rpr job, or at least this one: https://build.ros2.org/job/Rpr__launch__ubuntu_noble_amd64/259/testReport/junit/launch_xml.test.launch_xml/test_group/test_group/. The launch config corresponding to the test failure is created as `tuple('bar') -> ['BAR']` (launch config name resolved to `'bar'` -> value resolved to `'BAR'`) by this logic for `<group>`'s `keep` attributes/child entitites: https://github.com/ros2/launch/blob/e1d12d595f2a7af7341fd685ea53ad302ea49d60/launch/launch/actions/group_action.py#L104-L107

</p>
</details>

Reply:
Just to reuse `SomeSubstitutionsType` (or, rather, `SomeSubstitutionsType_types_tuple` in this case with `isinstance()`) in case it ever gets updated/extended again, we could probably do this for `is_substitutable()`:

```py
                from ..some_substitutions_type import SomeSubstitutionsType_types_tuple

                def is_substitutable(value: Any) -> bool:
                    # Specifically look for iterables of non-substitutable values. Assume the rest
                    # is substitutable and let it error out if not
                    return (
                        not isinstance(value, collections.abc.Iterable)
                        or all(
                            (
                                isinstance(iter_value, SomeSubstitutionsType_types_tuple)
                                and not isinstance(iter_value, collections.abc.Iterable)
                            )
                            for iter_value in value
                        )
                    )
```

This way we're not basically re-creating `SomeSubstitutionsType`/`SomeSubstitutionsType_types_tuple` here. Sorry for the liberal use of `()` to appease the linter :sweat_smile: